File: pascal/objects/pwm_libsimpleio.pas
{ PWM (Pulse Width Modulated) output services using libsimpleio               }

{ Copyright (C)2017-2018, Philip Munts, President, Munts AM Corp.             }
{                                                                             }
{ Redistribution and use in source and binary forms, with or without          }
{ modification, are permitted provided that the following conditions are met: }
{                                                                             }
{ * Redistributions of source code must retain the above copyright notice,    }
{   this list of conditions and the following disclaimer.                     }
{                                                                             }
{ THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" }
{ AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE   }
{ IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE  }
{ ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE   }
{ LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR         }
{ CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF        }
{ SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS    }
{ INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN     }
{ CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)     }
{ ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE  }
{ POSSIBILITY OF SUCH DAMAGE.                                                 }

UNIT PWM_libsimpleio;

INTERFACE

  USES
    libsimpleio,
    PWM;

  TYPE
    Polarities = (ActiveLow, ActiveHigh);

    OutputSubclass = CLASS(TInterfacedObject, PWM.Output)
      CONSTRUCTOR Create
       (output    : libsimpleio.Designator;
        frequency : Cardinal;
        dutycycle : Real = DUTYCYCLE_MIN;
        polarity  : Polarities = ActiveHigh);

      CONSTRUCTOR Create
       (chip      : Cardinal;
        channel   : Cardinal;
        frequency : Cardinal;
        dutycycle : Real = DUTYCYCLE_MIN;
        polarity  : Polarities = ActiveHigh);

      DESTRUCTOR Destroy; OVERRIDE;

      PROCEDURE WriteDutyCycle(dutycycle : Real);

      PROPERTY dutycycle : Real WRITE WriteDutyCycle;

    PRIVATE
      fd     : Integer;
      period : Integer;
    END;

IMPLEMENTATION

  USES
    errno,
    libPWM;

  { PWM_libsimpleio.OutputSubclass constructor }

  CONSTRUCTOR OutputSubclass.Create
   (output    : libsimpleio.Designator;
    frequency : Cardinal;
    dutycycle : Real;
    polarity  : Polarities);

  VAR
    ontime : Integer;
    error  : Integer;

  BEGIN
    IF (dutycycle < DUTYCYCLE_MIN) OR (dutycycle > DUTYCYCLE_MAX) THEN
      RAISE PWM.Error.Create('ERROR: Invalid duty cycle parameter, ' +
        errno.strerror(EINVAL));

    Self.period := Round(1.0E9/frequency);
    ontime := Round(dutycycle/DUTYCYCLE_MAX*period);

    libPWM.Configure(output.chip, output.chan, period, ontime,
      Ord(polarity), error);

    IF error <> 0 THEN
      RAISE PWM.Error.Create('ERROR: libPWM.Configure() failed, ' +
        errno.strerror(error));

    libPWM.Open(output.chip, output.chan, Self.fd, error);

    IF error <> 0 THEN
      RAISE PWM.Error.Create('ERROR: libPWM.Open() failed, ' +
        errno.strerror(error));
  END;

  CONSTRUCTOR OutputSubclass.Create
   (chip      : Cardinal;
    channel   : Cardinal;
    frequency : Cardinal;
    dutycycle : Real;
    polarity  : Polarities);

  VAR
    ontime : Integer;
    error  : Integer;

  BEGIN
    IF (dutycycle < DUTYCYCLE_MIN) OR (dutycycle > DUTYCYCLE_MAX) THEN
      RAISE PWM.Error.Create('ERROR: Invalid duty cycle parameter, ' +
        errno.strerror(EINVAL));

    Self.period := Round(1.0E9/frequency);
    ontime := Round(dutycycle/DUTYCYCLE_MAX*period);

    libPWM.Configure(chip, channel, period, ontime, Ord(polarity), error);

    IF error <> 0 THEN
      RAISE PWM.Error.Create('ERROR: libPWM.Configure() failed, ' +
        errno.strerror(error));

    libPWM.Open(chip, channel, Self.fd, error);

    IF error <> 0 THEN
      RAISE PWM.Error.Create('ERROR: libPWM.Open() failed, ' +
        errno.strerror(error));
  END;

  { PWM_libsimpleio.OutputSubclass destructor }

  DESTRUCTOR OutputSubclass.Destroy;

  VAR
    error  : Integer;

  BEGIN

    { Close the PWM output device file handle }

    libPWM.Close(Self.fd, error);

    IF error <> 0 THEN
      RAISE PWM.Error.Create('ERROR: libPWM.Close() failed, ' +
        errno.strerror(error));

    INHERITED;
  END;

  {  PWM_libsimpleio.OutputSubclass write method }

  PROCEDURE OutputSubclass.WriteDutyCycle(dutycycle : Real);

  VAR
    ontime : Integer;
    error  : Integer;

  BEGIN
    ontime := Round(dutycycle/DUTYCYCLE_MAX*Self.period);

    libPWM.Write(Self.fd, ontime, error);

    IF error <> 0 THEN
      RAISE PWM.Error.Create('ERROR: libPWM.Write() failed, ' +
        errno.strerror(error));
  END;

END.

File: c++/simpleio/pwm-libsimpleio.cpp
// PWM output services using libsimpleio

// Copyright (C)2017-2020, Philip Munts, President, Munts AM Corp.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
// * Redistributions of source code must retain the above copyright notice,
//   this list of conditions and the following disclaimer.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

#include <exception-raisers.h>
#include <pwm-libsimpleio.h>
#include <libsimpleio/libpwm.h>

using namespace libsimpleio::PWM;

// Constructor

Output_Class::Output_Class(unsigned chip, unsigned channel, unsigned frequency,
  double dutycycle, unsigned polarity)
{
  // Validate parameters

  if (frequency < 1)
    THROW_MSG("The frequency parameter is out of range");

  if (dutycycle < Interfaces::PWM::DUTYCYCLE_MIN)
    THROW_MSG("The dutycycle parameter is out of range");

  if (dutycycle > Interfaces::PWM::DUTYCYCLE_MAX)
    THROW_MSG("The dutycycle parameter is out of range");

  if (polarity > ActiveHigh)
    THROW_MSG("The polarity parameter is out of range");

  // Calculate the PWM pulse frequency and initial pulse width in nanoseconds

  unsigned period = 1.0E9/frequency;
  unsigned ontime = dutycycle/100.0*period;

  // Configure the PWM output

  int fd;
  int error;

  PWM_configure(chip, channel, period, ontime, polarity, &error);
  if (error) THROW_MSG_ERR("PWM_configure() failed", error);

  PWM_open(chip, channel, &fd, &error);
  if (error) THROW_MSG_ERR("PWM_open() failed", error);

  this->fd = fd;
  this->freq = frequency;
  this->period = period;
}

// PWM output methods

void Output_Class::write(const double dutycycle)
{
  // Validate parameters

  if (dutycycle < Interfaces::PWM::DUTYCYCLE_MIN)
    THROW_MSG("The dutycycle parameter is out of range");

  if (dutycycle > Interfaces::PWM::DUTYCYCLE_MAX)
    THROW_MSG("The dutycycle parameter is out of range");

  // Calculate the required PWM pulse width in nanoseconds

  unsigned ontime = dutycycle/100.0*this->period;
  int error;

  // Set the PWM pulse width

  PWM_write(this->fd, ontime, &error);
  if (error) THROW_MSG_ERR("PWM_write() failed", error);
}

// Query the configured PWM output pulse frequency

unsigned Output_Class::frequency(void)
{
  return this->freq;
}
